doc/developer/design/20210811_catalog_only_mode.md

```diff
@@ -0,0 +1,115 @@
+# Catalog-only mode
+
+## Summary
+
+When users have views causing OOM crashes, we want to provide a mechanism that
+still lets users boot into Materialize and take actions to remediate the issue,
+e.g. drop the OOMing view. To accomplish this, we'll introduce a new boot mode
+called "catalog-only mode" that starts Materialize as normal with the exception
+of not shipping any dataflows for user-created objects, such as materialized
+views.
+
+Ideally, we'll also provide a mechanism that lets users enable dataflows for
+their objects while Materialize is running using a statement such as `ALTER
+INDEX <index name> SET ENABLED`.
+
+## Goals
+
+- Support booting in catalog-only mode, which will not:
+    - Ship the dataflow for any user-created object
+    - Support querying user-created objects, such as views
+    - Support inserting into tables
+
+    Any other action should be allowed, such as creating, modifying, or dropping
+    objects.
+- Introduce a `--catalog-only` flag to Materialize that lets users easily turn
+  on catalog-only mode.
+- Add support for `ALTER INDEX ... SET ENABLED` to enable dataflows for objects
+  while Materialilze is running, and lift the restrictions imposed on
+  catalog-only mode for objects that use those dataflows.
+
+  Note that this statement will be "intrinsically idempotent," so `ALTER INDEX
+  ... SET ENABLED` will always succeed, even if Materialize was not booted in
+  catalog only mode or its dataflows are already active. This matches PG's
+  behavior for `ALTER` statements like `ALTER TABLE a ALTER COLUMN x SET NOT
+  NULL`, which you can run repeatedly.
+- Add no more than a single boolean check to the code path of any operation
+  occurring outside of catalog-only mode.
+
+## Non-Goals
+
+- We are not planning to add a statement to re-enable the dataflows for all
+  user-created objects; users can either accomplish this with a script or by
+  simply rebooting _not_ in catalog-only mode.
+
+## Description
+
+The `Coordinator` will handle the enforcing catalog-only mode on objects, i.e.
+disallowing dataflows from shipping, querying user-crated objects, or inserting
+into tables; these are the potentially prohibited actions. To support this enforcement,
+we'll add a new field to the `Coordinator`:
+
+```rust
+pub struct Coordinator {
+    ...
+    /// When `None`, all user objects can create/access dataflows; when `Some`, only the objects
+    /// with an ID in the hashmap are allowed.
+    catalog_only_mode_allow: Option<HashSet<GlobalId>>,
+}
+```
+
+When booting normally, we'll set this field to `None`, and
+`Some(HashSet::new())` when booting in catalog-only mode.
+
+Then, whenever a user or the system attempts to perform a potentially prohibited
```

Comment:
How do we identify potentially prohibited actions? What if we forget one and a user tries to run it? Will that produce a crash?

Is it possible to *not* add checks where needed, and instead make the dataflow side return errors for dataflows that don't exist?

Reply:
I looked into pushing some of these catches down into the dataflow layer, but I don't believe the dataflow layer and the coordinator have a relationship where that's immediately tractable. For instance, on `INSERT` for a table, there is simply a broadcast to perform the write, in the assumption that it's infallible. Someone with more insight into the dataflow layer might have a better idea if this is feasible through some other means.